cupyx/scipy/sparse/construct.py

```diff
@@ -88,6 +88,230 @@ def spdiags(data, diags, m, n, format=None):
+
+    .. seealso:: :func:`scipy.sparse.vstack`
+
+    Examples:
+        >>> from cupy.sparse import coo_matrix, vstack
+        >>> A = coo_matrix([[1, 2], [3, 4]])
+        >>> B = coo_matrix([[5, 6]])
+        >>> vstack([A, B]).toarray()
+        array([[1, 2],
+               [3, 4],
+               [5, 6]])
+    """
+    return bmat([[b] for b in blocks], format=format, dtype=dtype)
+
+
+def bmat(blocks, format=None, dtype=None):
+    """Builds a sparse matrix from sparse sub-blocks
+    Args:
+        blocks (array_like):
+            Grid of sparse matrices with compatible shapes.
+            An entry of None implies an all-zero matrix.
+        format ({'bsr', 'coo', 'csc', 'csr', 'dia', 'dok', 'lil'}, optional):
+            The sparse format of the result (e.g. "csr").  By default an
+            appropriate sparse matrix format is returned.
+            This choice is subject to change.
+        dtype (dtype, optional):
+            The data-type of the output matrix.  If not given, the dtype is
+            determined from that of `blocks`.
+    Returns:
+        bmat (sparse matrix)
+
+    .. seealso:: :func: `scipy.sparse.bmat`
+
+    Examples:
+        >>> from cupy.sparse import coo_matrix, bmat
+        >>> A = coo_matrix([[1, 2], [3, 4]])
+        >>> B = coo_matrix([[5], [6]])
+        >>> C = coo_matrix([[7]])
+        >>> bmat([[A, B], [None, C]]).toarray()
+        array([[1, 2, 5],
+               [3, 4, 6],
+               [0, 0, 7]])
+        >>> bmat([[A, None], [None, C]]).toarray()
+        array([[1, 2, 0],
+               [3, 4, 0],
+               [0, 0, 7]])
+    """
+
+    # We assume here that blocks will be 2-D so we need to look, at most,
+    # 2 layers deep for the shape
+
+    # NOTE: We can't follow scipy exactly here
+    # since we don't have an `object` datatype
+    M = len(blocks)
+    N = len(blocks[0])
+
+    blocks_flat = []
+    for m in range(M):
+        for n in range(N):
+            if blocks[m][n] is not None:
+                blocks_flat.append(blocks[m][n])
+
+    if len(blocks_flat) == 0:
+        return coo.coo_matrix((0, 0), dtype=dtype)
+
+    # if blocks.ndim != 2:
+    #     raise ValueError('blocks must be 2-D')
+
+    # check for fast path cases
+    if (N == 1 and format in (None, 'csr') and
+            all(isinstance(b, csr.csr_matrix)
+                for b in blocks_flat)):
+        A = _compressed_sparse_stack(blocks_flat, 0)
+        if dtype is not None:
+            A = A.astype(dtype)
+        return A
+    elif (M == 1 and format in (None, 'csc')
+          and all(isinstance(b, csc.csc_matrix) for b in blocks_flat)):
+        A = _compressed_sparse_stack(blocks_flat, 1)
+        if dtype is not None:
+            A = A.astype(dtype)
+        return A
+
+    block_mask = cupy.zeros((M, N), dtype=bool)
+    brow_lengths = cupy.zeros(M+1, dtype=cupy.int64)
+    bcol_lengths = cupy.zeros(N+1, dtype=cupy.int64)
```

Comment:
I think these don't have to be GPU arrays, because we have to loop over `M` and `N` on CPU anyway.
CPU workload won't be decreased by moving them on GPU.
How about using `numpy`? That way host-device synchronization can be removed.